### PR TITLE
Use `ar_archive_writer` for writing COFF import libs on all backends

### PR DESCRIPTION
**Pull request by ChrisDenton**
This is mostly the same as the llvm backend but with the cranelift version copy/pasted in place of the LLVM library.

try-job: x86_64-msvc
try-job: i686-msvc
try-job: i686-mingw
try-job: aarch64-gnu
try-job: aarch64-apple
try-job: test-various
try-job: armhf-gnu

### CODE REVIEW

File: compiler/rustc_codegen_llvm/Cargo.toml [hunk 1]
```diff
@@ -8,11 +8,15 @@ test = false
 
 [dependencies]
 # tidy-alphabetical-start
+ar_archive_writer = "0.4.0"
```
Comment (bjorn3) [inline compiler/rustc_codegen_llvm/Cargo.toml:11]: Adding this isn't necessary, right?

File: compiler/rustc_codegen_cranelift/src/archive.rs [hunk 2]
```diff
@@ -1,13 +1,6 @@
-use std::borrow::Borrow;
-use std::fs;
-use std::path::Path;
-
-use ar_archive_writer::{COFFShortExport, MachineTypes};
```
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/src/archive.rs:5]: Can you remove the `extern crate ar_archive_writer;` from `lib.rs` too?

File: compiler/rustc_codegen_llvm/Cargo.toml (hunk 1 above)
Comment (ChrisDenton) [reply compiler/rustc_codegen_llvm/Cargo.toml:11]: Oops, that was left over from my experimentations.

File: compiler/rustc_codegen_ssa/src/back/archive.rs [hunk 3]
```diff
@@ -34,7 +40,81 @@ pub trait ArchiveBuilderBuilder {
...
+                cpu => panic!("unsupported cpu type {cpu}"),
+            };
+
+            if let Err(error) = ar_archive_writer::write_import_library(
+                &mut file,
+                lib_name,
+                &exports,
+                machine,
+                !sess.target.is_like_msvc,
+                true,
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/archive.rs:111]: Maybe add a comment to this argument like
```suggestion
                /*comdat=*/ true,
```

File: compiler/rustc_codegen_llvm/Cargo.toml [hunk 4]
```diff
@@ -12,7 +12,10 @@ bitflags = "2.4.1"
 itertools = "0.12"
 libc = "0.2"
 measureme = "11"
-object = { version = "0.36.2", default-features = false, features = ["std", "read"] }
+object = { version = "0.36.2", default-features = false, features = [
+    "std",
+    "read",
+] }
```
Comment (bjorn3) [inline compiler/rustc_codegen_llvm/Cargo.toml:18]: Maybe revert the rest of the changes to this file too?

Comment (ChrisDenton) [reply compiler/rustc_codegen_llvm/Cargo.toml:18]: Sure. That's my IDE's auto formatter having fun again.

File: compiler/rustc_codegen_ssa/src/back/archive.rs [hunk 5]
```diff
@@ -34,7 +40,81 @@ pub trait ArchiveBuilderBuilder {
...
+                cpu => panic!("unsupported cpu type {cpu}"),
+            };
+
+            if let Err(error) = ar_archive_writer::write_import_library(
+                &mut file,
+                lib_name,
+                &exports,
+                machine,
+                !sess.target.is_like_msvc,
+                /*comdat=*/ true,
```
Comment (jieyouxu) [inline compiler/rustc_codegen_ssa/src/back/archive.rs:111]: Question: should this get a brief explanation / TL;DR for why comdat is used?

**Review by jieyouxu — APPROVED** [review]
This seems reasonable to me. Do we want any tests (even just a sanity check) for this (in terms of integration w/ ar_archive_writer)? Or does ar_archive_writer have sufficient tests?

**Comment by jieyouxu** [comment]
Since this PR would also close the linked issue, is it possible to have a regression test?

File: compiler/rustc_codegen_ssa/src/back/archive.rs (hunk 5 above)
Comment (ChrisDenton) [reply compiler/rustc_codegen_ssa/src/back/archive.rs:111]: Can do! And I can link to the issue for further explanation.

**Comment by ChrisDenton** [comment]
> Do we want any tests (even just a sanity check) for this (in terms of integration w/ ar_archive_writer)?

I think this should be covered by our existing raw-dylib tests. They should fail if anything goes wrong.

> Since this PR would also close the linked issue, is it possible to have a regression test?

I can try.

File: tests/run-make/msvc-wholearchive/rmake.rs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+// Reason: this is testing the MSVC linker
+
+// This is a regression test for #129020
+// It ensures we can link a rust staticlib into DLL using the MSVC linker
+
+use run_make_support::run::cmd;
+use run_make_support::rustc;
+
+fn main() {
+    // BUild the staticlib
```
Comment (bjorn3) [inline tests/run-make/msvc-wholearchive/rmake.rs:11]: *Build (casing for second letter is wrong)

**Review by jieyouxu — APPROVED** [review]
Thanks!

**Comment by ChrisDenton** [comment]
Failed in rollup https://github.com/rust-lang/rust/pull/129198#issuecomment-2294877985

@bors r-

**Comment by ChrisDenton** [comment]
While I figure out what went wrong, let's make sure x86_64 passes

@bors try

**Comment by jieyouxu** [comment]
... weird

**Comment by ChrisDenton** [comment]
Hm, it seems to work on x86_64 with MSVC's `link.exe` but it seems there's problems with LLVM and i686. It might be that the need the `.idata$3` object file to be just so. It might still be possible to workaround that limitation but not easily.

So I think I'll edit this to disable the comdat option and just land the unification of COFF writing for now. It might be a good idea to land them separately anyway.

**Comment by jieyouxu** [comment]
Thanks!

@bors r+

**Pull request merged**